### PR TITLE
Merge consecutive apertures

### PR DESCRIPTION
**Pull request by freddieknets**
## Description

1. Added a function `merge_consecutive_apertures()` to the `line` to collapse consecutive apertures: in the case of 3 or more consecutive apertures (only separated by `Marker`s or `Drift`s) that are exactly the same, all are removed except the first and the last
2. Expanded the other simplification methods (`remove_zero_length_drifts()`, `merge_consecutive_drifts()`,  `remove_markers()`, `remove_inactive_multipoles()`, and `merge_consecutive_multipoles()`): added a `keep` option to all, and implemented the `inplace=False` functionality
3. Added two tests for `merge_consecutive_apertures()` 
4. Expanded the existing tests for the other simplification methods, to take the new options for `keep` and `inplace` into account
5. Created two helper functions to the `line`: `_apertures_equal()` to check if two apertures are the same, and `_lines_equal()` which compares two lines for equality (this currently only works if all elements in the line have a `to_dict()` method, and raises a `NotImplementedError` otherwise)
6. Updated `check_aperture()` to use the `_is_aperture()` and `_allow_backtrack` helper functions. This improves code consistency, and avoids the need to update the code when new elements are added (like the `XRotation` and `YRotation`)

## Checklist

Mandatory: 

- [X] I have added tests to cover my changes
- [X] All the tests are passing, including my new ones
- [X] I described my changes in this PR description

Optional:

- [X] The code I wrote follows good style practices (see [PEP 8](https://peps.python.org/pep-0008/) and [PEP 20](https://peps.python.org/pep-0020/)).
- [ ] I have updated the docs in relation to my changes, if applicable
- [ ] I have tested also GPU contexts

### CODE REVIEW

File: xtrack/line.py [hunk 1]
```diff
@@ -726,101 +768,177 @@ def get_s_position(self, at_elements=None, mode="upstream"):
...
-            names.append(nn)
+            if isinstance(ee, Marker) and nn not in keep:
+                continue
+            newline.append_element(ee, nn)
 
-        self.element_names = names
-        return self
+        if inplace:
+            self.element_names = newline.element_names
+            self.element_dict  = newline.element_dict
```
Comment (giadarol) [inline xtrack/line.py:788]: I think rebinding element_dict breaks the xdeps expressions

File: xtrack/line.py [hunk 2]
```diff
@@ -726,101 +768,177 @@ def get_s_position(self, at_elements=None, mode="upstream"):
...
                 ctx2np = ee._context.nparray_from_context_array
                 aux = ([ee.hxl, ee.hyl]
                         + list(ctx2np(ee.knl)) + list(ctx2np(ee.ksl)))
                 if np.sum(np.abs(np.array(aux))) == 0.0:
                     continue
             newline.append_element(ee, nn)
 
+        if inplace:
+            self.element_names = newline.element_names
+            self.element_dict  = newline.element_dict
```
Comment (giadarol) [inline xtrack/line.py:814]: I think rebinding element_dict breaks the xdeps expressions

File: xtrack/line.py [hunk 3]
```diff
@@ -726,101 +768,177 @@ def get_s_position(self, at_elements=None, mode="upstream"):
...
 
-            if _is_drift(ee):
+            if _is_drift(ee) and not nn in keep:
                 prev_nn = newline.element_names[-1]
                 prev_ee = newline.element_dict[prev_nn]
                 if _is_drift(prev_ee):
                     prev_ee.length += ee.length
                 else:
-                    newline.append_element(ee, nn)
+                    newline.append_element(ee.copy(), nn)
```
Comment (giadarol) [inline xtrack/line.py:863]: Why do you make a copy?

File: xtrack/line.py (hunk 1 above)
Comment (rdemaria) [reply xtrack/line.py:788]: xdeps updates values in `_owner` attribute of the ref. It is possible to replace it to a new dictionary if needed.

Comment (freddieknets) [reply xtrack/line.py:810]: I reinitialised the ref; that should do it I believe.

**Comment by freddieknets** [comment]
I made a few fixes:
1. `ee.copy()` is only done in case `inplace=False` (not needed otherwise)
2. var management is correctly initialised / copied in the simplification methods
3. var management is tested for equality in `_lines_equal`
4. the lines used in the simplification tests now have a var management, so that  this is properly tested.

All tests pass on CPU

**Comment by giadarol** [comment]
For some reason it seems to compare to an old version. I'll close and reopen

**Comment by giadarol** [comment]
Followed up in #313

**Pull request closed without merging**